RIOT-OS/RIOT
# pkg/semtech-loramac: rework interaction with the MAC

### Contribution description

This is another attempt to fix #11530.
It follows the suggestion of @ParksProjets in https://github.com/RIOT-OS/RIOT/pull/11535#issuecomment-493151047:
- TX status is now only reported after an mcps confirm event
- RX data are notified from mcps indication events
- one needs an external thread to receive data from the MAC. This is much simpler but it consume a little more RAM if one wants to manage downlinks message outside of the MAC. The bonus point with this approach is the ability to correctly support class C devices.

I have tested this PR using OTAA, ABP activations and with CONFIRMED and UNCONFIRMED messages.

Everything is working as expected.

The only small thing is the link check which is not working correctly but it's already the case in master I think.

Finally, I also had to update the `examples/lorawan` application and doxygen documentation to avoid potential lock of autonomous application. Maybe `examples/lorawan` could be later refactored with downling messages better handled or adapted for class C devices.




### Testing procedure

Run `tests/pkg_semtech-loramac`, send messages from the device, send messages from the network using CONFIRMED/UNCONFIRMED (`cnf/uncnf` option of `loramac tx`: everything should work, with this PR, the application doesn't block.




### Issues/PRs references

fixes #11530, closes #11535 (cleaner alternative)
